In [1]:
import pandas as pd

df = pd.read_csv('/content/4) house Prediction Data Set.csv')
display(df.head())
print(df.info())
display(df.describe())

0.00632  18.00   2.310  0  0.5380  6.5750  65.20  4.0900   1  296.0  15.30 396.90   4.98  24.00
0   0.02731   0.00   7.070  0  0.4690  6.4210  78...                                             
1   0.02729   0.00   7.070  0  0.4690  7.1850  61...                                             
2   0.03237   0.00   2.180  0  0.4580  6.9980  45...                                             
3   0.06905   0.00   2.180  0  0.4580  7.1470  54...                                             
4   0.02985   0.00   2.180  0  0.4580  6.4300  58...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 505 entries, 0 to 504
Data columns (total 1 columns):
 #   Column                                                                                            Non-Null Count  Dtype 
---  ------                                                                                            --------------  ----- 
 0    0.00632  18.00   2.310  0  0.5380  6.5750  65.20  4.0900   1  296.0  15.30 396.90   4.98  24.00  505 non-null    object
dtypes: object(1)
memory usage: 4.1+ KB
None


0.00632  18.00   2.310  0  0.5380  6.5750  65.20  4.0900   1  296.0  15.30 396.90   4.98  24.00
count                                                 505                                             
unique                                                505                                             
top      0.04741   0.00  11.930  0  0.5730  6.0300  80...                                             
freq                                                    1

In [2]:
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
df = pd.read_csv('/content/4) house Prediction Data Set.csv', header=None, delimiter=r"\s+", names=column_names)
display(df.head())
print(df.info())
display(df.describe())

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB
None


CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

In [3]:
print(df.isnull().sum())

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64


In [4]:
print(df.dtypes)

CRIM       float64
ZN         float64
INDUS      float64
CHAS         int64
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD          int64
TAX        float64
PTRATIO    float64
B          float64
LSTAT      float64
MEDV       float64
dtype: object


In [5]:
from sklearn.model_selection import train_test_split

X = df.drop('MEDV', axis=1)
y = df['MEDV']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Training set shape:", X_train.shape, y_train.shape)
print("Testing set shape:", X_test.shape, y_test.shape)

Training set shape: (404, 13) (404,)
Testing set shape: (102, 13) (102,)


In [16]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error

# Create a Linear Regression model instance
linear_model = LinearRegression()

# Train the model
linear_model.fit(X_train, y_train)

# Make predictions on the training and test sets
y_train_pred_linear = linear_model.predict(X_train)
y_test_pred_linear = linear_model.predict(X_test)

# Calculate evaluation metrics for training set
mse_linear_train = mean_squared_error(y_train, y_train_pred_linear)
r2_linear_train = r2_score(y_train, y_train_pred_linear)

# Calculate evaluation metrics for test set
mse_linear_test = mean_squared_error(y_test, y_test_pred_linear)
r2_linear_test = r2_score(y_test, y_test_pred_linear)

# Print the metrics
print("Linear Regression Performance:")
print(f"Train MSE: {mse_linear_train:.2f}, R-squared: {r2_linear_train:.2f}")
print(f"Test MSE: {mse_linear_test:.2f}, R-squared: {r2_linear_test:.2f}")

Linear Regression Performance:
Train MSE: 21.64, R-squared: 0.75
Test MSE: 24.29, R-squared: 0.67


In [7]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [17]:
# Create and train the Decision Tree Regressor model
decision_tree_model = DecisionTreeRegressor(random_state=42)
decision_tree_model.fit(X_train, y_train)

# Make predictions on the training and test sets
y_train_pred_tree = decision_tree_model.predict(X_train)
y_test_pred_tree = decision_tree_model.predict(X_test)

# Calculate evaluation metrics for the Decision Tree model
mse_tree_train = mean_squared_error(y_train, y_train_pred_tree)
r2_tree_train = r2_score(y_train, y_train_pred_tree)
mse_tree_test = mean_squared_error(y_test, y_test_pred_tree)
r2_tree_test = r2_score(y_test, y_test_pred_tree)

# Print the metrics for the Decision Tree model
print("Decision Tree Regressor Performance:")
print(f"Train MSE: {mse_tree_train:.2f}, R-squared: {r2_tree_train:.2f}")
print(f"Test MSE: {mse_tree_test:.2f}, R-squared: {r2_tree_test:.2f}")

Decision Tree Regressor Performance:
Train MSE: 0.00, R-squared: 1.00
Test MSE: 10.42, R-squared: 0.86


In [18]:
# Create and train the Random Forest Regressor model
random_forest_model = RandomForestRegressor(random_state=42, n_estimators=100)
random_forest_model.fit(X_train, y_train)

# Make predictions on the training and test sets
y_train_pred_forest = random_forest_model.predict(X_train)
y_test_pred_forest = random_forest_model.predict(X_test)

# Calculate evaluation metrics for the Random Forest model
mse_forest_train = mean_squared_error(y_train, y_train_pred_forest)
r2_forest_train = r2_score(y_train, y_train_pred_forest)
mse_forest_test = mean_squared_error(y_test, y_test_pred_forest)
r2_forest_test = r2_score(y_test, y_test_pred_forest)

# Print the metrics for the Random Forest model
print("Random Forest Regressor Performance:")
print(f"Train MSE: {mse_forest_train:.2f}, R-squared: {r2_forest_train:.2f}")
print(f"Test MSE: {mse_forest_test:.2f}, R-squared: {r2_forest_test:.2f}")

Random Forest Regressor Performance:
Train MSE: 1.98, R-squared: 0.98
Test MSE: 7.90, R-squared: 0.89


In [10]:
# Create a dictionary to store the results
results = {
    'Model': ['Linear Regression', 'Decision Tree', 'Random Forest'],
    'MSE': [mse_linear, mse_tree, mse_forest],
    'MAE': [mae_linear, mae_tree, mae_forest],
    'R-squared': [r2_linear, r2_tree, r2_forest]
}

# Create a DataFrame from the results dictionary
results_df = pd.DataFrame(results)

# Print the comparison table
print("Model Performance Comparison:")
display(results_df)

Model Performance Comparison:


Model        MSE       MAE  R-squared
0  Linear Regression  24.291119  3.189092   0.668759
1      Decision Tree  10.416078  2.394118   0.857963
2      Random Forest   7.901514  2.039539   0.892253

In [11]:
from sklearn.model_selection import GridSearchCV

In [12]:
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
}

grid_search = GridSearchCV(estimator=RandomForestRegressor(random_state=42),
                           param_grid=param_grid,
                           cv=5,
                           n_jobs=-1,
                           scoring='neg_mean_squared_error')

grid_search.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=RandomForestRegressor(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [None, 10, 20, 30],
                         'n_estimators': [100, 200, 300]},
             scoring='neg_mean_squared_error')

In [13]:
print("Best parameters found: ", grid_search.best_params_)
print("Best score (neg_mean_squared_error): ", grid_search.best_score_)

Best parameters found:  {'max_depth': None, 'n_estimators': 100}
Best score (neg_mean_squared_error):  -14.858836289320982


In [19]:
best_forest = grid_search.best_estimator_
y_pred_best_forest = best_forest.predict(X_test)

mse_best_forest = mean_squared_error(y_test, y_pred_best_forest)
mae_best_forest = mean_absolute_error(y_test, y_pred_best_forest)
r2_best_forest = r2_score(y_test, y_pred_best_forest)

print("Optimized Random Forest Performance:")
print(f"MSE: {mse_best_forest:.2f}")
print(f"MAE: {mae_best_forest:.2f}")
print(f"R-squared: {r2_best_forest:.2f}")

# Add the new results to the comparison DataFrame
new_results = {
    'Model': 'Optimized Random Forest',
    'MSE': mse_best_forest,
    'MAE': mae_best_forest,
    'R-squared': r2_best_forest
}
new_results_df = pd.DataFrame([new_results])
results_df = pd.concat([results_df, new_results_df], ignore_index=True)


print("\nUpdated Model Performance Comparison:")
display(results_df)

Optimized Random Forest Performance:
MSE: 7.90
MAE: 2.04
R-squared: 0.89

Updated Model Performance Comparison:


Model        MSE       MAE  R-squared
0        Linear Regression  24.291119  3.189092   0.668759
1            Decision Tree  10.416078  2.394118   0.857963
2            Random Forest   7.901514  2.039539   0.892253
3  Optimized Random Forest   7.901514  2.039539   0.892253
4  Optimized Random Forest   7.901514  2.039539   0.892253

In [15]:
best_forest = grid_search.best_estimator_
y_pred_best_forest = best_forest.predict(X_test)

mse_best_forest = mean_squared_error(y_test, y_pred_best_forest)
mae_best_forest = mean_absolute_error(y_test, y_pred_best_forest)
r2_best_forest = r2_score(y_test, y_pred_best_forest)

print("Optimized Random Forest Performance:")
print(f"MSE: {mse_best_forest:.2f}")
print(f"MAE: {mae_best_forest:.2f}")
print(f"R-squared: {r2_best_forest:.2f}")

# Add the new results to the comparison DataFrame
new_results_df = pd.DataFrame([new_results])
results_df = pd.concat([results_df, new_results_df], ignore_index=True)

print("\nUpdated Model Performance Comparison:")
display(results_df)

Optimized Random Forest Performance:
MSE: 7.90
MAE: 2.04
R-squared: 0.89

Updated Model Performance Comparison:


Model        MSE       MAE  R-squared
0        Linear Regression  24.291119  3.189092   0.668759
1            Decision Tree  10.416078  2.394118   0.857963
2            Random Forest   7.901514  2.039539   0.892253
3  Optimized Random Forest   7.901514  2.039539   0.892253

# Executive Summary

## Summary of Findings

*   **Best Performing Model:** The **Random Forest Regressor** demonstrated the highest accuracy in predicting house prices, outperforming both Linear Regression and Decision Tree models.
*   **Model Performance:** The Random Forest model achieved an **R-squared of 0.89** on the test data, indicating that it can explain 89% of the variance in house prices.
*   **Data Quality:** The dataset was clean with no missing values, which streamlined the model-building process.
*   **Hyperparameter Tuning:** The GridSearchCV process did not yield a significant improvement in model performance, suggesting that the default parameters of the Random Forest model were already robust for this dataset.

## Recommendations

*   **Model Adoption:** The current **Random Forest model** is recommended for use in predicting house prices due to its strong performance and reliability.
*   **Explore Advanced Models:** For potentially higher accuracy, it is recommended to experiment with more advanced models such as **XGBoost** or **LightGBM**.
*   **Feature Engineering:** Further improvements could be achieved by exploring feature engineering techniques to create new, more informative features from the existing data.